### PR TITLE
feat: jumpstart model artifact instance type variants

### PR DESCRIPTION
**Pull request by evakravi**
*Issue #, if available:*

*Description of changes:*
JumpStart plans on introducing models that may have different model artifacts depending on the instance type chosen. This will require new metadata, and new SDK code for consuming this metadata.

*Testing done:*
Unit tests added

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/jumpstart/types.py [hunk 1]
```diff
@@ -346,6 +346,62 @@ def to_json(self) -> Dict[str, Any]:
...
 
+    def get_instance_specific_prepacked_artifact_key(self, instance_type: str) -> Optional[str]:
+        """Returns instance specific model artifact key.
+
+        Returns None if a model, instance type tuple does not have specific
+        artifact key.
+        """
+
+        return self._get_instance_specific_property(
+            instance_type=instance_type, property_name="prepacked_artifact_key"
```
Comment (whittech1) [inline src/sagemaker/jumpstart/types.py:357]: nit: should the property name include `s3`?

File: src/sagemaker/jumpstart/artifacts/model_uris.py [hunk 2]
```diff
@@ -84,14 +87,55 @@ def _retrieve_model_uri(
...
 
+    model_artifact_key: str
+
     if model_scope == JumpStartScriptScope.INFERENCE:
+        instance_specific_prepacked_hosting_artifact_key: Optional[str] = (
+            model_specs.hosting_instance_type_variants.get_instance_specific_prepacked_artifact_key(
+                instance_type=instance_type
+            )
+            if instance_type
+            and getattr(model_specs, "hosting_instance_type_variants", None) is not None
```
Comment (whittech1) [inline src/sagemaker/jumpstart/artifacts/model_uris.py:98]: Isn't the name of the property `hosting_custom_instance_type_variants` in the MH PR?

File: src/sagemaker/jumpstart/artifacts/model_uris.py [hunk 3]
```diff
@@ -84,14 +87,55 @@ def _retrieve_model_uri(
...
+        default_prepacked_hosting_artifact_key: Optional[str] = getattr(
+            model_specs, "hosting_prepacked_artifact_key"
+        )
+
+        default_hosting_artifact_key: str = model_specs.hosting_artifact_key
+
         model_artifact_key = (
-            getattr(model_specs, "hosting_prepacked_artifact_key", None)
-            or model_specs.hosting_artifact_key
+            instance_specific_prepacked_hosting_artifact_key
```
Comment (whittech1) [inline src/sagemaker/jumpstart/artifacts/model_uris.py:118]: I think this logic can break if there is an instance-specific normal artifact but a default prepacked artifact. Previously, just the presence of default prepacked artifact would have indicated a need to prepack.

To be exact, we need to change the execution flow based on whether `model_specs` is prepacked or not. Quick-and-dirty approach would be something like:

```
is_prepacked = model_specs.hosting_use_script_uri or default_prepacked_hosting_artifact_key
if is_prepacked:
    return instance_specific_prepack_key or default_prepacked_hosting_artifact_key
else:
    return instance_specific_key or default_hosting_artifact_key
```

File: src/sagemaker/jumpstart/artifacts/model_uris.py [hunk 4]
```diff
@@ -50,6 +51,8 @@ def _retrieve_model_uri(
             artifact S3 URI.
         model_scope (str): The model type, i.e. what it is used for.
             Valid values: "training" and "inference".
+        instance_type (str): An instance type to optionally supply in order to get
+            model artifacts specific for the instance type. (Default: None).
```
Comment (JGuinegagne) [inline src/sagemaker/jumpstart/artifacts/model_uris.py:55]: `Optional. The ML compute instance type for the specified scope. (Default: None).`

File: src/sagemaker/jumpstart/artifacts/model_uris.py [hunk 5]
```diff
@@ -84,14 +87,55 @@ def _retrieve_model_uri(
...
+        default_hosting_artifact_key: str = model_specs.hosting_artifact_key
+
         model_artifact_key = (
-            getattr(model_specs, "hosting_prepacked_artifact_key", None)
-            or model_specs.hosting_artifact_key
+            instance_specific_prepacked_hosting_artifact_key
+            or instance_specific_hosting_artifact_key
+            or default_prepacked_hosting_artifact_key
+            or default_hosting_artifact_key
         )
```
Comment (JGuinegagne) [inline src/sagemaker/jumpstart/artifacts/model_uris.py:122]: The hosting artifact retrieval logic is quite complex and the implementation is too linear IMO. Could you break it down into utilities that you can unit test separately.

E.g.
```py
if model_scope == JumpStartScriptScope.INFERENCE:
  if is_model_prepacked(...):
    return artifacts.retrieve_instance_specific_hosting_prepack_key(...)
  else:
    return artifacts.retrieve_instance_specific_hosting_artifact_key(...)
elif model_scope === JumpStartScriptScope.TRAINING:
  return artifacts.retrieve_instance_specific_training_artifact_key(...)
```

Then delegate to each utility method the task of handling whether (a) an instance was passed, (b) if (a) is true, whether it maps + add unit tests for each case.

File: src/sagemaker/model_uris.py [hunk 6]
```diff
@@ -44,6 +45,8 @@ def retrieve(
             the model artifact S3 URI.
         model_scope (str): The model type.
             Valid values: "training" and "inference".
+        instance_type (str): An instance type to optionally supply in order to get
+            model artifacts specific for the instance type. (Default: None).
```
Comment (JGuinegagne) [inline src/sagemaker/model_uris.py:49]: same comment here.

File: tests/unit/sagemaker/jumpstart/constants.py [hunk 7]
```diff
@@ -373,6 +373,232 @@
...
+                "p3": {"regional_properties": {"image_uri": "$gpu_image_uri"}},
+                "p4": {"regional_properties": {"image_uri": "$gpu_image_uri"}},
+                "g4dn": {"regional_properties": {"image_uri": "$gpu_image_uri"}},
+                "m2": {"regional_properties": {"image_uri": "$cpu_image_uri"}},
+                "c2": {"regional_properties": {"image_uri": "$cpu_image_uri"}},
+                "ml.g5.48xlarge": {
+                    "properties": {"environment_variables": {"TENSOR_PARALLEL_DEGREE": "8"}}
+                },
+                "ml.g5.12xlarge": {
+                    "properties": {"environment_variables": {"TENSOR_PARALLEL_DEGREE": "4"}}
```
Comment (JGuinegagne) [inline tests/unit/sagemaker/jumpstart/constants.py:410]: could you add a test with an instance specific `prepacked_artifact_key` as well please?

As in not just instance family variant please.

Same for training case please

**Review by JGuinegagne — CHANGES_REQUESTED** [review]
Could you add specific unit tests for (a) `sagemaker/model_uri` and (b) `ssagemaker/jumpstart/artifacts/model_uris` please?

I know you are implicitely testing through `JumpStartModel` and `JumpStartEstimator`, but we should guard ourselves against specific logic introduced in these classes and test the utilities directly as well.

**Review by JGuinegagne — COMMENTED** [review]
Could you update the docstring as well please?

**Review by whittech1 — APPROVED** [review]

**Review by evakravi — COMMENTED** [review]
/bot run slow-tests

**Review by evakravi — COMMENTED** [review]
/bot run pr

**Comment by knikure** [comment]
@evakravi Can you resolve conflicts on this PR?

**Pull request merged**